File: server/src/models/Job.ts
import mongoose, { Document, Schema } from 'mongoose';

export type JobType = 'full-time' | 'part-time' | 'internship' | 'contract' | 'remote';
export type JobStatus = 'open' | 'closed';

export interface IJob extends Document {
  _id: mongoose.Types.ObjectId;
  title: string;
  company: string;
  location: string;
  description: string;
  requirements: string[];
  salary?: string;
  type: JobType;
  postedBy: mongoose.Types.ObjectId;
  source: 'internal';
  status: JobStatus;
  applicationCount: number;
  createdAt: Date;
  updatedAt: Date;
}

const JobSchema = new Schema<IJob>(
  {
    title: {
      type: String,
      required: [true, 'Job title is required'],
      trim: true,
      maxlength: [150, 'Title cannot exceed 150 characters'],
    },
    company: {
      type: String,
      required: [true, 'Company name is required'],
      trim: true,
    },
    location: {
      type: String,
      required: [true, 'Location is required'],
      trim: true,
    },
    description: {
      type: String,
      required: [true, 'Job description is required'],
    },
    requirements: {
      type: [String],
      default: [],
    },
    salary: { type: String },
    type: {
      type: String,
      enum: ['full-time', 'part-time', 'internship', 'contract', 'remote'],
      required: [true, 'Job type is required'],
    },
    postedBy: {
      type: Schema.Types.ObjectId,
      ref: 'User',
      required: true,
    },
    source: {
      type: String,
      default: 'internal',
    },
    status: {
      type: String,
      enum: ['open', 'closed'],
      default: 'open',
    },
    applicationCount: {
      type: Number,
      default: 0,
    },
  },
  { timestamps: true }
);

JobSchema.index({ title: 'text', company: 'text', location: 'text', description: 'text' });

export default mongoose.model<IJob>('Job', JobSchema);

File: server/app/models/job.py
from beanie import Document, Link, Indexed
from pydantic import Field
from typing import Optional, Literal, List
from bson import ObjectId
from datetime import datetime

JobType = Literal["full-time", "part-time", "internship", "contract", "remote"]
JobStatus = Literal["open", "closed"]


class Job(Document):
    title: str = Field(..., max_length=150)
    company: str
    location: str
    description: str
    requirements: List[str] = Field(default_factory=list)
    salary: Optional[str] = None
    type: JobType
    posted_by: str = Field(..., alias="postedBy")   # stored as string ObjectId
    source: str = Field(default="internal")
    status: JobStatus = Field(default="open")
    application_count: int = Field(default=0, alias="applicationCount")
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)

    class Settings:
        name = "jobs"

    model_config = {"populate_by_name": True}

    def to_dict(self, posted_by_info: Optional[dict] = None) -> dict:
        data = self.model_dump(by_alias=True)
        data["_id"] = str(self.id)
        if posted_by_info:
            data["postedBy"] = posted_by_info
        return data
